tests/server_test.php

```diff
@@ -0,0 +1,142 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Loaded REST.
+ *
+ * @package webservice_loadedrest
+ * @author Luke Carrier <luke@carrier.im>
+ * @copyright 2018 Luke Carrier
+ */
+
+use org\bovigo\vfs\vfsStream;
+use org\bovigo\vfs\vfsStreamFile;
+use webservice_loadedrest\format\format;
+use webservice_loadedrest\server;
+
+defined('MOODLE_INTERNAL') || die;
+
+class webservice_loadedrest_server_mock extends server {
+    protected function authenticate_by_token($tokentype) {
+        global $USER;
+
+        $this->restricted_context = context_system::instance();
+
+        return $USER;
+    }
+
+    protected function parse_request() {
+        parent::parse_request();
+        $this->functionname = 'webservice_loadedrest_external_mock_function';
+    }
+
+    protected function load_function_info() {
+        $function = (object) [
+            'id' => 1000000,
+            'name' => $this->functionname,
+            'classname' => 'webservice_loadedrest_external_mock',
+            'methodname' => 'function',
+            'classpath' => __FILE__,
+            'component' => 'webservice_loadedrest',
+            'capabilities' => '',
+            'services' => null,
+        ];
+        $this->function = external_api::external_function_info($function);
+    }
+}
+
+class webservice_loadedrest_external_mock extends external_api {
+    public static function function_parameters() {
+        return new external_function_parameters([
+            'vfsStream' => new external_value(
+                    PARAM_TEXT, '', VALUE_REQUIRED),
+        ]);
+    }
+
+    public static function function_returns() {
+        return new external_value(
+                PARAM_BOOL, '', VALUE_REQUIRED);
+    }
+
+    public static function function($params) {}
```

Comment:
This is meant to mirror the upsert() method in `auth/avadosalesforce/classes/external/upsert_v1.php`, right? But do we need to define it here inside this class? And if we do, can it use a more telling name, as 'function' is pretty generic?

Reply:
Yes in the sense that it's the implementation of a WS function, but no as they're separate plugins -- this is just a mock. If you have a better name I'll take it, but I'm keen to follow the naming convention in the [external functions API](https://docs.moodle.org/dev/External_functions_API).

Reply:
I'd go with `mock_upsert_parameters`, `mock_upsert_returns` and `mock_upsert`, but it's not a life-and-death issue.

Reply:
I'll meet you halfway with `mock_external_function`, since this is _not_ directly related to `auth_avadosalesforce`.